resourcex/extract.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcex
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+type ExtractOptions struct {
+	// TagName is the struct tag name to use for resource property names; defaults to "json".
+	TagName string
+	// RejectUnknowns produces an error (of type ContainsUnknownsError) if any unknowns are extracted.
+	RejectUnknowns bool
+}
+
+type ExtractResult struct {
+	// Dependencies is the set of resources that the extracted information depends on (known or unknown).
+	Dependencies []resource.URN
```

Comment:
I don't think we need to keep ordering here. I can't think of a situation where relying on the order of dependencies is correct.
```suggestion
	Dependencies map[resource.URN]struct{}
```